File: netDxf/Entities/PolyfaceMeshRecord.cs
using System;
using System.Collections.Generic;
using System.Linq;
using netDxf.Header;
using netDxf.IO;
using netDxf.Tables;

namespace netDxf.Entities
{
    /// <summary>A retained VERTEX or SEQEND belonging to a loaded polyface mesh.</summary>
    /// <remarks>
    /// Geometry remains in PolyfaceMesh.Vertexes. These records expose identity and common metadata;
    /// they are not independently insertable entities. Stored optional fields and private packets
    /// survive without interpreting private application data. The initial implementation is source bound.
    /// </remarks>
    public sealed class PolyfaceMeshRecord : DxfObject
    {
        internal readonly List<DxfTag> Tags;
        internal readonly Dictionary<int, DxfObject> Resources = new Dictionary<int, DxfObject>();
        internal readonly Dictionary<int, string> OriginalResourceNames = new Dictionary<int, string>();
        internal readonly Dictionary<short, int> Coordinates = new Dictionary<short, int>();
        internal readonly Dictionary<int, int> MetadataGroups = new Dictionary<int, int>();
        internal readonly List<string> ReactorHandles = new List<string>();
        internal readonly List<DxfObject> OriginalReactors = new List<DxfObject>();
        internal string ExtensionHandle;
        internal DxfObject OriginalExtension;
        internal string SourceOwner;
        internal int CommonEnd;
        internal int XDataStart;
        internal Vector3 Position;
        internal DxfDocument SourceDocument;
        internal int IdentityIndex;
        internal int OwnerIndex;
        internal bool HasPrivateData;
        internal int CoordinateIndex = -1;
        internal int FaceIndex = -1;
        internal readonly Dictionary<short, int> FaceSlots = new Dictionary<short, int>();
        internal short[] OriginalIndexes;
        internal PolyfaceMeshFace StoredFace;
        internal AciColor OriginalColor;
        internal int FaceLayerIndex = -1;
        internal int FaceCommonEnd = -1;
        internal readonly HashSet<int> FaceColorIndices = new HashSet<int>();
        /// <summary>Gets whether this VERTEX describes a face through signed coordinate indices.</summary>
        public bool IsFaceRecord { get; internal set; }
        /// <summary>Gets the existing face model for a face record, or null for a coordinate or terminator.</summary>
        public PolyfaceMeshFace Face { get { return this.StoredFace; } }
        internal bool FaceColorUnchanged
        { get { return this.Face == null || this.Face.Color == null && this.OriginalColor == null
            || this.Face.Color != null && this.OriginalColor != null && this.Face.Color.Index == this.OriginalColor.Index
                && this.Face.Color.UseTrueColor == this.OriginalColor.UseTrueColor && this.Face.Color.Equals(this.OriginalColor); } }
        internal bool FaceIndexesUnchanged { get { return this.Face == null || this.Face.VertexIndexes.SequenceEqual(this.OriginalIndexes); } }

        internal PolyfaceMeshRecord(string codeName, List<DxfTag> tags) : base(codeName)
        { this.Tags = tags; this.XDataStart = tags.Count; }

        /// <summary>Gets the DXF profile in which this stored record was loaded.</summary>
        /// <remarks>Stored optional and private packets require the same profile on output or adoption.</remarks>
        public DxfVersion SourceVersion { get; internal set; }
        /// <summary>Gets the stored layer identity, or null when group 8 was absent.</summary>
        public Layer Layer { get { return this.Face != null ? this.Face.Layer : this.Resources.Values.OfType<Layer>().FirstOrDefault(); } }
        /// <summary>Gets the stored linetype identity, or null when group 6 was absent.</summary>
        public Linetype Linetype { get { return this.Resources.Values.OfType<Linetype>().FirstOrDefault(); } }
        /// <summary>Gets whether this record terminates the vertex sequence.</summary>
        public bool IsSequenceEnd { get { return this.CodeName == DxfObjectCode.EndSequence; } }
        /// <summary>Gets whether the stored VERTEX group 330 names the containing block record.</summary>
        /// <remarks>The structural Owner is always the containing PolyfaceMesh. This observed producer form is retained independently.</remarks>
        public bool UsesBlockRecordOwner { get; internal set; }
        /// <summary>Gets the current identity written in ordinary group 330.</summary>
        /// <remarks>For the block-record form this follows an explicit move of the containing polyface mesh to another block.</remarks>
        public DxfObject StoredOwner
        { get { return this.UsesBlockRecordOwner ? ((this.Owner as PolyfaceMesh)?.Owner as netDxf.Blocks.Block)?.Record : this.Owner; } }
        internal IEnumerable<DxfObject> References
        { get { return this.Face == null ? this.Resources.Values : this.Resources.Values.Where(item => !(item is Layer))
                .Concat(this.Face.Layer == null ? Enumerable.Empty<DxfObject>() : new DxfObject[] { this.Face.Layer }); } }
        internal IEnumerable<DxfTag> OpaqueHandleTags
        {
            get
            {
                for (int i = 0; i < this.XDataStart; i++)
                {
                    if (this.MetadataGroups.TryGetValue(i, out int last)) { i = last; continue; }
                    if (i != this.IdentityIndex && i != this.OwnerIndex && !this.Resources.ContainsKey(i)
                        && netDxf.Objects.DxfObjectDatabase.IsReference(this.Tags[i])) yield return this.Tags[i];
                }
            }
        }

        internal void Validate(DxfDocument document, PolyfaceMesh owner, bool registered)
        {
            if (document.DrawingVariables.AcadVer != this.SourceVersion)
                throw new NotSupportedException("Converting retained VERTEX/SEQEND records to another DXF profile requires schema regeneration.");
            if (this.SourceDocument != null && !ReferenceEquals(this.SourceDocument, document) || !ReferenceEquals(this.Owner, owner))
                throw new InvalidOperationException("A retained polyface mesh record must stay with its source polyface mesh and document.");
            DxfObject current = this.Handle == null ? null : document.GetObjectByHandle(this.Handle);
            if (registered ? !ReferenceEquals(current, this) : current != null && !ReferenceEquals(current, this))
                throw new InvalidOperationException("A retained polyface mesh record has inconsistent registration.");
            if (registered && (this.StoredOwner == null || !ReferenceEquals(document.GetObjectByHandle(this.StoredOwner.Handle), this.StoredOwner)))
                throw new InvalidOperationException("A retained polyface mesh record has an invalid stored owner.");
            if (this.SourceDocument == null && !registered) return;
            foreach (DxfObject target in this.References)
                if (!ReferenceEquals(document.GetObjectByHandle(target.Handle), target))
                    throw new InvalidOperationException("A retained polyface mesh record resource is outside its source document.");
        }
        internal bool CanClone()
        {
            return !this.HasPrivateData && this.ExtensionDictionary == null && this.PersistentReactors.Count == 0
                && this.Resources.Values.All(resource => resource is Layer || resource is Linetype)
                && !this.XData.Values.SelectMany(data => data.XDataRecord).Any(tag => tag.Code == XDataCode.DatabaseHandle
                    && ulong.Parse((string)tag.Value, System.Globalization.NumberStyles.AllowHexSpecifier, System.Globalization.CultureInfo.InvariantCulture) != 0);
        }
        internal int TopologyTagCount()
        {
            int count = this.XDataStart;
            if (this.Face != null)
            {
                count += (this.Face.Layer == null ? 0 : 1) - (this.FaceLayerIndex < 0 ? 0 : 1);
                if (!this.FaceColorUnchanged)
                    count += (this.Face.Color == null ? 0 : this.Face.Color.UseTrueColor ? 2 : 1) - this.FaceColorIndices.Count;
            }
            bool extension = false, reactors = false;
            foreach (var group in this.MetadataGroups)
            {
                bool isExtension = (string)this.Tags[group.Key].Value == "{ACAD_XDICTIONARY";
                bool unchanged = isExtension ? ReferenceEquals(this.ExtensionDictionary, this.OriginalExtension)
                    : this.PersistentReactors.SequenceEqual(this.OriginalReactors)
                        && this.ReactorHandles.Where(handle => handle != "0").SequenceEqual(this.PersistentReactors.Select(target => target.Handle));
                if (!unchanged)
                {
                    count -= group.Value - group.Key + 1;
                    count += isExtension ? this.ExtensionDictionary == null ? 0 : 3
                        : this.PersistentReactors.Count == 0 ? 0 : this.PersistentReactors.Count + 2;
                }
                if (isExtension) extension = true; else reactors = true;
            }
            if (!extension && this.ExtensionDictionary != null) count += 3;
            if (!reactors && this.PersistentReactors.Count != 0) count += this.PersistentReactors.Count + 2;
            foreach (XData data in this.XData.Values) count += 1 + data.XDataRecord.Count;
            return count;
        }
        internal PolyfaceMeshRecord CopyForClone(PolyfaceMeshFace face = null)
        {
            var result = new PolyfaceMeshRecord(this.CodeName, new List<DxfTag>(this.Tags))
            {
                SourceOwner = this.SourceOwner, CommonEnd = this.CommonEnd, XDataStart = this.XDataStart,
                Position = this.Position, IdentityIndex = this.IdentityIndex, OwnerIndex = this.OwnerIndex,
                ExtensionHandle = this.ExtensionHandle, UsesBlockRecordOwner = this.UsesBlockRecordOwner, SourceVersion = this.SourceVersion,
                CoordinateIndex = this.CoordinateIndex, FaceIndex = this.FaceIndex, IsFaceRecord = this.IsFaceRecord,
                FaceLayerIndex = this.FaceLayerIndex, FaceCommonEnd = this.FaceCommonEnd,
                OriginalIndexes = this.OriginalIndexes == null ? null : (short[])this.OriginalIndexes.Clone(), StoredFace = face,
                OriginalColor = this.OriginalColor == null ? null : (AciColor)this.OriginalColor.Clone()
            };
            foreach (int index in this.FaceColorIndices) result.FaceColorIndices.Add(index);
            foreach (var pair in this.FaceSlots) result.FaceSlots.Add(pair.Key, pair.Value);
            foreach (var pair in this.Coordinates) result.Coordinates.Add(pair.Key, pair.Value);
            foreach (var pair in this.MetadataGroups) result.MetadataGroups.Add(pair.Key, pair.Value);
            result.ReactorHandles.AddRange(this.ReactorHandles);
            foreach (var pair in this.OriginalResourceNames) result.OriginalResourceNames.Add(pair.Key, pair.Value);
            foreach (var pair in this.Resources)
            {
                if (pair.Value is Layer layer)
                {
                    Layer current = face == null ? (Layer)layer.Clone() : face.Layer;
                    if (current != null) result.Resources.Add(pair.Key, current);
                }
                else result.Resources.Add(pair.Key, (Linetype)((Linetype)pair.Value).Clone());
            }
            foreach (XData data in this.XData.Values) result.XData.Add((XData)data.Clone());
            return result;
        }
    }
}

File: javascript/netDxf/Entities/PolyfaceMeshRecord.js
// Copyright (c) netDxf contributors. MIT License; see package LICENSE.
import { DxfObject } from '../DxfObject.js';
import { DxfObjectCode } from '../DxfObjectCode.js';
import { Vector3 } from '../Vector3.js';
import { Layer } from '../Tables/Layer.js';
import { Linetype } from '../Tables/Linetype.js';
import { Copy } from '../../runtime/GeometryRuntime.js';
import { ReferenceList } from '../../runtime/ReferenceList.js';
import { FixedArray } from '../../runtime/FixedArray.js';
import { IsStoredReference, HasNonzeroXDataReference, SameSequence } from '../../runtime/StoredRecord.js';
import { InvalidCastException, InvalidOperationException, NotSupportedException } from '../../runtime/Errors.js';

/** Original internal-constructor record model. Creating this does not register a DXF record. */
export class PolyfaceMeshRecord extends DxfObject {
  Tags; Resources = new Map(); OriginalResourceNames = new Map(); Coordinates = new Map(); MetadataGroups = new Map();
  ReactorHandles = new ReferenceList(); OriginalReactors = new ReferenceList(); ExtensionHandle = null; OriginalExtension = null;
  SourceOwner = null; CommonEnd = 0; XDataStart; Position = Vector3.Zero; SourceDocument = null;
  IdentityIndex = 0; OwnerIndex = 0; HasPrivateData = false; CoordinateIndex = -1; FaceIndex = -1;
  FaceSlots = new Map(); OriginalIndexes = null; StoredFace = null; OriginalColor = null;
  FaceLayerIndex = -1; FaceCommonEnd = -1; FaceColorIndices = new Set();
  IsFaceRecord = false; SourceVersion = 0; UsesBlockRecordOwner = false;
  constructor(codeName, tags) { super(codeName); this.Tags = tags; this.XDataStart = tags.Count ?? tags.length; }
  get Face() { return this.StoredFace; }
  get FaceColorUnchanged() {
    const face = this.Face, original = this.OriginalColor;
    return face == null || (face.Color == null && original == null) ||
      (face.Color != null && original != null && face.Color.Index === original.Index && face.Color.UseTrueColor === original.UseTrueColor && face.Color.Equals(original));
  }
  get FaceIndexesUnchanged() { return this.Face == null || SameSequence(this.Face.VertexIndexes, this.OriginalIndexes); }
  get Layer() { return this.Face != null ? this.Face.Layer : Array.from(this.Resources.values()).find(v => v instanceof Layer) ?? null; }
  get Linetype() { return Array.from(this.Resources.values()).find(v => v instanceof Linetype) ?? null; }
  get IsSequenceEnd() { return this.CodeName === DxfObjectCode.EndSequence; }
  get StoredOwner() {
    if (!this.UsesBlockRecordOwner) return this.Owner;
    const block = this.Owner?.Owner;
    return block?.CodeName === DxfObjectCode.Block ? block.Record ?? null : null;
  }
  get References() {
    const resources = Array.from(this.Resources.values());
    return this.Face == null ? resources : resources.filter(v => !(v instanceof Layer)).concat(this.Face.Layer == null ? [] : [this.Face.Layer]);
  }
  get OpaqueHandleTags() {
    const record = this;
    return { *[Symbol.iterator]() {
      for (let i = 0; i < record.XDataStart; i++) {
        if (record.MetadataGroups.has(i)) { i = record.MetadataGroups.get(i); continue; }
        const tag = record.Tags.get_Item ? record.Tags.get_Item(i) : record.Tags[i];
        if (i !== record.IdentityIndex && i !== record.OwnerIndex && !record.Resources.has(i) && IsStoredReference(tag)) yield tag;
      }
    } };
  }
  Validate(document, owner, registered) {
    if (document.DrawingVariables.AcadVer !== this.SourceVersion) throw new NotSupportedException('Retained record version conversion requires schema regeneration.');
    if ((this.SourceDocument != null && this.SourceDocument !== document) || this.Owner !== owner) throw new InvalidOperationException('Retained polyface source ownership changed.');
    const current = this.Handle == null ? null : document.GetObjectByHandle(this.Handle);
    if (registered ? current !== this : current != null && current !== this) throw new InvalidOperationException('Inconsistent retained record registration.');
    if (registered && (this.StoredOwner == null || document.GetObjectByHandle(this.StoredOwner.Handle) !== this.StoredOwner)) throw new InvalidOperationException('Invalid retained stored owner.');
    if (this.SourceDocument == null && !registered) return;
    for (const target of this.References) if (document.GetObjectByHandle(target.Handle) !== target) throw new InvalidOperationException('Retained resource is outside its source document.');
  }
  CanClone() {
    return !this.HasPrivateData && this.ExtensionDictionary == null && this.PersistentReactors.Count === 0 &&
      Array.from(this.Resources.values()).every(r => r instanceof Layer || r instanceof Linetype) && !HasNonzeroXDataReference(this);
  }
  TopologyTagCount() {
    let count = this.XDataStart;
    if (this.Face != null) {
      count += (this.Face.Layer == null ? 0 : 1) - (this.FaceLayerIndex < 0 ? 0 : 1);
      if (!this.FaceColorUnchanged) count += (this.Face.Color == null ? 0 : this.Face.Color.UseTrueColor ? 2 : 1) - this.FaceColorIndices.size;
    }
    let extension = false, reactors = false;
    for (const [first, last] of this.MetadataGroups) {
      const tag = this.Tags.get_Item ? this.Tags.get_Item(first) : this.Tags[first];
      const isExtension = tag.Value === '{ACAD_XDICTIONARY';
      const unchanged = isExtension ? this.ExtensionDictionary === this.OriginalExtension :
        SameSequence(this.PersistentReactors, this.OriginalReactors) && SameSequence(Array.from(this.ReactorHandles).filter(h => h !== '0'), Array.from(this.PersistentReactors, t => t.Handle));
      if (!unchanged) {
        count -= last - first + 1;
        count += isExtension ? (this.ExtensionDictionary == null ? 0 : 3) : (this.PersistentReactors.Count === 0 ? 0 : this.PersistentReactors.Count + 2);
      }
      if (isExtension) extension = true; else reactors = true;
    }
    if (!extension && this.ExtensionDictionary != null) count += 3;
    if (!reactors && this.PersistentReactors.Count !== 0) count += this.PersistentReactors.Count + 2;
    for (const data of this.XData.Values) count += 1 + data.XDataRecord.Count;
    return count;
  }
  CopyForClone(face = null) {
    const copy = new PolyfaceMeshRecord(this.CodeName, new ReferenceList(this.Tags));
    for (const name of ['SourceOwner','CommonEnd','XDataStart','Position','IdentityIndex','OwnerIndex','ExtensionHandle','UsesBlockRecordOwner','SourceVersion','CoordinateIndex','FaceIndex','IsFaceRecord','FaceLayerIndex','FaceCommonEnd']) copy[name] = Copy(this[name]);
    copy.OriginalIndexes = this.OriginalIndexes == null ? null : FixedArray(this.OriginalIndexes);
    copy.StoredFace = face; copy.OriginalColor = this.OriginalColor == null ? null : this.OriginalColor.Clone();
    for (const index of this.FaceColorIndices) copy.FaceColorIndices.add(index);
    for (const name of ['FaceSlots','Coordinates','MetadataGroups','OriginalResourceNames']) for (const [key, value] of this[name]) copy[name].set(key, value);
    copy.ReactorHandles.AddRange(this.ReactorHandles);
    for (const [key, value] of this.Resources) {
      if (value instanceof Layer) {
        const layer = face == null ? value.Clone() : face.Layer;
        if (layer != null) copy.Resources.set(key, layer);
      } else {
        if (!(value instanceof Linetype)) throw new InvalidCastException();
        copy.Resources.set(key, value.Clone());
      }
    }
    for (const data of this.XData.Values) copy.XData.Add(data.Clone());
    return copy;
  }
}
